### ISSUE TITLE
Xcode 12.5 and iOS 14.5 "Error: The provided scheme is not valid" [with solution]

### ISSUE THREAD

**Issue by mjsarfatti** [AyogoHealth/cordova-plugin-oauth#18]
Testing version 3.0.1 of the plugin on iOS 14.5 I get the error:

```
[AuthenticationSession] Error: The provided scheme is not valid. A scheme must not include any special characters, such as ":" or "/"
```

which I can track back to `OAuthPlugin.swift`.

It appears to be the same issue as in https://github.com/auth0/react-native-auth0/issues/360, which is that "Since version 12.5 (released on 26 Apr 2021), Xcode no longer accepts an entire URL for the ASWebAuthenticationSession scheme value."

That issue has a merged PR: https://github.com/auth0/react-native-auth0/pull/369

While I don't know Swift nor C++, I was able to somehow translate that PR and implement a fix that _appears_ to be working.

I say "appears" because due to operating in a strict enterprise development environment I can test only half of the OAuth flow, which is also the reason I'm not sending a PR.

The fix is simply adding two lines (and renaming the corresponding parameter):

```swift
    required init(_ endpoint : URL, callbackScheme : String) {
        var url: URL = URL(string: callbackScheme)!                     ⬅️
        var callbackURLScheme: String = url.scheme ?? callbackScheme    ⬅️               ⬇️
        self.aswas = ASWebAuthenticationSession(url: endpoint, callbackURLScheme: callbackURLScheme, completionHandler: { (callBack:URL?, error:Error?) in
            if let incomingUrl = callBack {
                NotificationCenter.default.post(name: NSNotification.Name.CDVPluginHandleOpenURL, object: incomingUrl)
            }
        })
    }
```

Again, I can't test the whole flow, so if someone could apply that fix and test appropriately would be great!

**Comment by flaresys** [comment]
The fix mentioned above worked for me on iOS 14.7.1 devices that were experiencing this issue.

**Comment by HarelM** [comment]
I currently see that when trying to use window.open on iOS (I tested it on an iPad) the app crashes.
Was this the behavior that you experienced?
I'll try to take the fix you mention and apply it and see what happens in any case...

**Comment by HarelM** [comment]
I've added the relevant fix in PR #23 if anyone is interested, seems to solve the crash that is reported here.

**Comment by dpogue** [comment]
I've released v3.0.2 and v2.0.2 with this fix.
